js-messaging/js-messaging-impl/src/main/java/com/duckduckgo/js/messaging/impl/RealJsMessageHandler.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.js.messaging.impl
+
+import android.webkit.WebView
+import androidx.annotation.MainThread
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.global.plugins.PluginPoint
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.js.messaging.api.JsMessage
+import com.duckduckgo.js.messaging.api.JsMessageHandler
+import com.duckduckgo.js.messaging.api.JsMessageHandlerPlugin
+import com.duckduckgo.js.messaging.api.JsRequestResponse
+import com.duckduckgo.js.messaging.api.JsRequestResponse.Error
+import com.duckduckgo.js.messaging.api.JsRequestResponse.Success
+import com.duckduckgo.js.messaging.api.SubscriptionEvent
+import com.squareup.anvil.annotations.ContributesBinding
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.Moshi
+import dagger.SingleInstanceIn
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import logcat.logcat
+
+@SingleInstanceIn(AppScope::class)
+@ContributesBinding(AppScope::class)
+class RealJsMessageHandler @Inject constructor(
+    private val handlers: PluginPoint<JsMessageHandlerPlugin>,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
+) : JsMessageHandler {
+
+    private val moshi = Moshi.Builder().add(JSONObjectAdapter()).build()
+    override fun process(message: String, context: String, secret: String, callback: String, webView: WebView) {
+        try {
+            val adapter = moshi.adapter(JsMessage::class.java)
+            val jsMessage = adapter.fromJson(message)
+
+            jsMessage?.let {
+                if (context == jsMessage.context) {
```

Comment:
can we lift the checf for` jsMessage.id != null` here?

Reply:
The check should be done in each plugin. We have two types of messages that we can received, notifications (without id) and requests (with id). Only the latter require a response from us so is up to the plugin to decide if it should return a response or not and you can't build a JsResponse object without id.